### PR TITLE
Fix NPE on duplicate field in record declaration

### LINKED ISSUE #20259: Compiler fails with NPE
**Description:**
Compiler fails with NPE on attached code.

**Steps to reproduce:**
Compile attached code

**Affected Versions:**
Ballerina 1.0.5

**OS, DB, other environment details and versions:**
Windows 10, JDK jdk1.8.0_231, IntelliJ

[Ballerina.toml.txt](https://github.com/ballerina-platform/ballerina-lang/files/3935584/Ballerina.toml.txt)

[listener.bal.txt](https://github.com/ballerina-platform/ballerina-lang/files/3935581/listener.bal.txt)
[ballerina-internal.log](https://github.com/ballerina-platform/ballerina-lang/files/3935582/ballerina-internal.log)

### PR DESCRIPTION
**Pull request by irshadnilam**
## Purpose
$title

Fixes #20259

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.ballerinalang.test.util.BAssertUtil;
+import org.ballerinalang.test.util.BCompileUtil;
+import org.ballerinalang.test.util.CompileResult;
+import org.testng.annotations.Test;
+
+/**
+ * Test cases for user record definition negative test.
+ */
+public class RecordDefNegativeTest {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: Do we need a separate class for this? Can't we just add these cases to https://github.com/ballerina-platform/ballerina-lang/blob/v1.1.0/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/OpenRecordNegativeTest.java#L81

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Test(description = "Negative test cases for record definition")
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: The description does not reflect what's being tested IMO.

File: tests/jballerina-unit-test/src/test/resources/test-src/record/negative/duplicate-key-negative.bal [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+     };
+}
+
+function duplicateKeyTest2() {
+    Song hi;
+    hi = {
+            name: "hi",
+            length: 2,
+            name: "hi"
+        };
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/negative/duplicate-key-negative.bal]: Shall we fix formatting?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java (hunk 1 above)
Comment (irshadnilam) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: This test case doesn't just belong to OpenRecord (or CloseRecords). Hence I thought it'd make sense this way.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: But we seem to be having tests for the same thing in two different places, which is not right IMO. Can we at least move those tests here?

Comment (irshadnilam) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: +1 will move those here

Comment (irshadnilam) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/record/RecordDefNegativeTest.java]: I have removed the mentioned duplicate-key tests, as it is tested from the tests I have added.

**Comment by irshadnilam** [comment]
> Are we planning on doing these changes on the 1.1.x branch also? IMO, we should.

@MaryamZi 
Please refer to the following PR to 1.1.x branch.
https://github.com/ballerina-platform/ballerina-lang/pull/20514

**Comment by irshadnilam** [comment]
The changes made after @KavinduZoysa approval were just rebasing to squash multiple commit messages.

**Pull request merged**